### PR TITLE
WT-9805 Save the updates need to be deleted from history store and delete them later

### PR DESCRIPTION
**Pull request by quchenhao**

### CODE REVIEW

File: src/include/reconcile.h
Comment (etienneptl) [inline src/include/reconcile.h:235]: Should we call it `delete_hs_upd_size` or `delete_hs_upd_count`?

File: src/history/hs_rec.c
Comment (etienneptl) [inline src/history/hs_rec.c:1067]: Can we add this comment:
```
/* Ensure we can see all the content in the history store. */
```

File: src/history/hs_rec.c
Comment (etienneptl) [inline src/history/hs_rec.c:1077]: Instead of the `WT_ASSERT`, should we do the same `if` as on line 1070? This race condition seems to be a problem though. If we have `ret == WT_NOTFOUND` but more content has been inserted to the hs just then, the assert will be false?

File: src/history/hs_rec.c
Comment (etienneptl) [inline src/history/hs_rec.c:1104]: In the case of an error, we would jump to the label `err` and clear the flag of `hs_cursor` if required. Why wouldn't we clear the flag(s) for the `tombstone` and `update` as well?
Instead of:
```
WT_ERR(r->hs_cursor->remove(r->hs_cursor));
```
We could simply do:
```
ret = r->hs_cursor->remove(r->hs_cursor);
```
What do you think?

File: src/include/btmem.h
Comment (etienneptl) [inline src/include/btmem.h:251]: I would say "Update **that** needs ...". Or "Information related to an update that should be removed from the history store".

File: src/include/reconcile.h
Comment (etienneptl) [inline src/include/reconcile.h:231]: I would reword with:
"List of history store updates to be deleted. While reviewing updates for each page, we save them here and delete them after we have built the disk image.".

File: src/reconcile/rec_visibility.c
Comment (etienneptl) [inline src/reconcile/rec_visibility.c:44]: `WT_DELETE_HS_UPDATE` -> `WT_DELETE_HS_UPD`

Comment (etienneptl) [reply src/reconcile/rec_visibility.c:44]: Could we say:
"Save an update into a WT_DELETE_HS_UPDATE list"? I am not sure if it's necessary to say that it is an update that needs to be deleted from the HS later. This function only serves the purpose of adding some content to a `WT_DELETE_HS_UPD`.

File: src/history/hs_rec.c
Comment (etienneptl) [inline src/history/hs_rec.c:1052]: Can we reword to:
"Delete an update from the history store"? We can probably add information about the case when we don't have to delete anything.

File: src/history/hs_rec.c
Comment (etienneptl) [inline src/history/hs_rec.c:1121]: Do you think we should check for the number of elements to be deleted before making this call?
it would be something like this:
```
if(r->delete_hs_upd_next == 0) goto done;
WT_ERR(__wt_scr_alloc(session, WT_INTPACK64_MAXSIZE, &key));
```

File: src/include/reconcile.h
Comment (quchenhao) [reply src/include/reconcile.h:235]: The name is consistent to other similar use cases. I intend to keep the same pattern.

File: src/history/hs_rec.c
Comment (quchenhao) [reply src/history/hs_rec.c:1077]: No, we already have a cursor position, inserting more things will not impact the current cursor position.

File: src/history/hs_rec.c
Comment (quchenhao) [reply src/history/hs_rec.c:1104]: If we failed to remove, then the update is still in the history store, we cannot clear the flag in this case. We should retry to remove again next time.

File: src/history/hs_rec.c
Comment (quchenhao) [reply src/history/hs_rec.c:1121]: yes, we can do that. What we gain is not allocating the key buffer in this case.

File: src/history/hs_rec.c
Comment (quchenhao) [reply src/history/hs_rec.c:1077]: No, it cannot happen, the assert is on tombstone. Doesn't have anything to do with newly inserted content.

File: src/reconcile/rec_visibility.c
Comment (quchenhao) [reply src/reconcile/rec_visibility.c:44]: I would still keep the history store here to make the intention clear.

**Review by etienneptl — APPROVED** [review]
LGTM!
Good job @quchenhao, would be great to update de description section of the PR though with a quick summary of the changes.

File: src/history/hs_rec.c
Comment (quchenhao) [reply src/history/hs_rec.c:63]: This is still in this file but I moved it to the end of the file.

File: src/history/hs_rec.c
Comment (quchenhao) [reply src/history/hs_rec.c:444]: yes, that's correct.

File: src/include/btmem.h
Comment (quchenhao) [reply src/include/btmem.h:258]: I put it here because __wt_save_upd is here. May be we should move both of them to reoncile.h

Comment (quchenhao) [reply src/include/btmem.h:258]: __wt_save_upd is referred in multi block so it is here. I think we can move __wt_delete_hs_upd to reconcile.h

File: src/btree/bt_cursor.c
Comment (quchenhao) [reply src/btree/bt_cursor.c:617]: The code is OK to not clear it. There is no correctness issue. I did this because I see in the core dump that we have garbage value for the values so I think better to fix that to not confuse us when debugging.

File: src/history/hs_rec.c
Comment (quchenhao) [reply src/history/hs_rec.c:352]: done

File: src/include/btmem.h
Comment (quchenhao) [reply src/include/btmem.h]: done

File: src/history/hs_rec.c
Comment (agorrod) [inline src/history/hs_rec.c:1127]: I think this can use WT_RET - if it returns an error, the scratch buffer won't have been allocated, so doesn't need to be cleaned up.

File: src/history/hs_rec.c
Comment (agorrod) [inline src/history/hs_rec.c:1135]: Did you base this code off other code? I can't see it being handled prior to this change - was there just a bug in how column store content was cleaned up?
Could you get around calculating it by saving a recno in the delete_hs_upd structure?

Comment (quchenhao) [reply src/history/hs_rec.c:1135]: I referred to the code in __wt_hs_insert_updates. By saving ins and rip we can pack the key for all the cases for column store and row store. If we choose to save recno, we still need to save ins and rip as there is case that row store needs ins if it only has an insert list.

File: src/history/hs_rec.c
Comment (quchenhao) [reply src/history/hs_rec.c]: done

File: src/history/hs_rec.c
Comment (agorrod) [inline src/history/hs_rec.c:1138]: The code this came from was using `WT_WITH_BTREE` - should this code do that as well? Is it possible that without the right btree, the resolution will use the history store tree?

File: src/history/hs_rec.c
Comment (agorrod) [reply src/history/hs_rec.c:1134]: Got it - I'd rather not have two copies of this code. It doesn't look like it would be obvious to factor it out into a shared function, but could you take a look?

File: src/history/hs_rec.c
Comment (quchenhao) [reply src/history/hs_rec.c:1138]: Good point, I'll fix that.

Comment (quchenhao) [reply src/history/hs_rec.c]: done

**Review by agorrod — APPROVED** [review]
Thanks @quchenhao - I'm happy with this now. It definitely requires a full review from someone else as well.

I've added in stress testing to the PR - I think w should wait for that to complete before merging.

**Review by kommiharibabu — APPROVED** [review]
LGTM

**Comment by quchenhao** [comment]
evergreen merge

**Pull request merged**
